nucliadb/nucliadb/search/search/paragraphs.py

```diff
@@ -0,0 +1,301 @@
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+import asyncio
+import logging
+import re
+import string
+from typing import Optional
+
+from redis import asyncio as aioredis
+from redis.asyncio.client import Redis
+
+from nucliadb.ingest.fields.base import Field, FieldTypes
+from nucliadb.ingest.orm.resource import KB_REVERSE
+from nucliadb.ingest.orm.resource import Resource as ResourceORM
+from nucliadb.search.settings import settings
+from nucliadb_telemetry import metrics
+from nucliadb_utils.utilities import get_utility, has_feature, set_utility
+
+from .cache import get_resource_from_cache
+
+logger = logging.getLogger(__name__)
+PRE_WORD = string.punctuation + " "
+
+CACHE_OPS = metrics.Counter("paragraph_cache_ops", labels={"type": "miss"})
+CACHE_HIT_DISTRIBUTION = metrics.Histogram(
+    "paragraph_cache_dist", buckets=[1, 2, 4, 8, 16, 32, 64, 128, 512, 1024]
+)
+
+_PARAGRAPHS_CACHE_UTIL = "paragraphs_cache"
+
+
+class ParagraphsCache:
+    """
+    Skeleton of paragraph cache.
+
+    For now, it will be used for us to track hits/misses on a potential
+    paragraph cache implementation.
+    """
+
+    consumer_task: Optional[asyncio.Task] = None
+    redis: Redis
+
+    def __init__(self):
+        self.queue = asyncio.Queue()
+
+    async def initialize(self) -> None:
+        if (
+            settings.search_cache_redis_host is None
+            or settings.search_cache_redis_port is None
+        ):
+            # Cache is not configured, ignore
+            return
+        self.consumer_task = asyncio.create_task(self.queue_consumer())
+        self.redis = aioredis.from_url(
+            f"redis://{settings.search_cache_redis_host}:{settings.search_cache_redis_port}"
+        )
+
+    async def finalize(self) -> None:
+        if self.consumer_task is None:
+            return
+        self.consumer_task.cancel()
+        await self.redis.close(close_connection_pool=True)
+
+    async def queue_consumer(self) -> None:
+        while True:
+            try:
+                key = await self.queue.get()
+                key = key + "_hits"
+                val = await self.redis.get(key)
+                if val is None:
+                    CACHE_OPS.inc({"type": "miss"})
+                else:
+                    CACHE_OPS.inc({"type": "hit"})
+                    CACHE_HIT_DISTRIBUTION.observe(int(val))
+                await self.redis.incr(key, 1)
+                await self.redis.expire(key, 60 * 10)
```

Comment:
mh, maybe this value be bigger? since we don't have that much traffic yet...

Reply:
it's tough to balance because we don't want something too large and then we always get hits.